### PR TITLE
Syntax for defining custom errors.

### PR DESCRIPTION
**Pull request by chriseth**
Split out of https://github.com/ethereum/solidity/pull/10922

Added the check for reserved signatures, but I'm still looking for an example for the all-ones-signature.

### CODE REVIEW

File: libsolidity/analysis/ContractLevelChecker.cpp [hunk 1]
```diff
@@ -433,6 +433,25 @@ void ContractLevelChecker::checkHashCollisions(ContractDefinition const& _contra
 			);
 		hashes.insert(hash);
 	}
+
+	map<uint32_t, SourceLocation> errorHashes;
+	// TODO all used errors?
```
Comment (hrkrshnn) [inline libsolidity/analysis/ContractLevelChecker.cpp]: Will this be done after the call graph PR?

File: libsolidity/ast/AST.h [hunk 2]
```diff
@@ -509,6 +509,7 @@ class ContractDefinition: public Declaration, public StructurallyDocumented, pub
 	std::vector<FunctionDefinition const*> definedFunctions() const { return filteredNodes<FunctionDefinition>(m_subNodes); }
 	std::vector<EventDefinition const*> events() const { return filteredNodes<EventDefinition>(m_subNodes); }
 	std::vector<EventDefinition const*> const& interfaceEvents() const;
+	std::vector<ErrorDefinition const*> errors() const { return filteredNodes<ErrorDefinition>(m_subNodes); }
```
Comment (hrkrshnn) [inline libsolidity/ast/AST.h]: Shouldn't we have the equivalent of `interfaceEvents()` for NatSpec and ABI?

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 3]
```diff
@@ -473,6 +473,18 @@ bool ASTJsonConverter::visit(EventDefinition const& _node)
 	return false;
 }
 
+bool ASTJsonConverter::visit(ErrorDefinition const& _node)
+{
+	m_inEvent = true;
```
Comment (hrkrshnn) [inline libsolidity/ast/ASTJsonConverter.cpp]:
```suggestion

```

File: libsolidity/ast/ASTUtils.cpp [hunk 4]
```diff
@@ -67,4 +67,14 @@ VariableDeclaration const* rootConstVariableDeclaration(VariableDeclaration cons
 	return rootDecl;
 }
 
+Declaration const* referencedDeclaration(Expression const& _expression)
```
Comment (hrkrshnn) [inline libsolidity/ast/ASTUtils.cpp]: We could remove lines 31 to 38 in the same file.

**Review by hrkrshnn — COMMENTED** [review]
Tests are failing.

File: libsolidity/analysis/ContractLevelChecker.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp]: yes

File: libsolidity/ast/AST.h (hunk 2 above)
Comment (chriseth) [reply libsolidity/ast/AST.h]: Will be done after the callgraph PR.

Comment (chriseth) [reply libsolidity/ast/AST.h]: I added a function that will be extended once we have the call graph.

**Comment by hrkrshnn** [comment]
`test266151307()` has the selector `0xffffffff`. Only took two minutes on a single thread to compute it!

File: test/libsolidity/syntaxTests/errors/weird2.sol [hunk 5]
```diff
@@ -0,0 +1,7 @@
+error E();
+
+contract C {
+    bytes4 t = E.selector;
```
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/errors/weird2.sol]: Does it have to do with the fact that `E` is internal?

**Comment by chriseth** [comment]
Ah great, thanks!
Added example and also changed the formatting of the selector to hex.

File: test/libsolidity/syntaxTests/errors/error_location_memory.sol [hunk 6]
```diff
@@ -0,0 +1,3 @@
+error E(uint[] memory);
+// ----
+// ParserError 2314: (15-21): Expected ',' but got 'memory'
```
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/errors/error_location_memory.sol:3]: This error message is rather confusing. Not sure if improving this is easy.

File: libsolidity/parsing/Parser.cpp [hunk 7]
```diff
@@ -112,8 +112,16 @@ ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
 				nodes.push_back(parseFunctionDefinition(true));
 				break;
 			default:
+				if (
+					// Workaround because `error` is not a keyword.
```
Comment (hrkrshnn) [inline libsolidity/parsing/Parser.cpp:116]: Will we make it a keyword in 0.9.0?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 8]
```diff
@@ -908,6 +908,10 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 			m_context << logInstruction(numIndexed);
 			break;
 		}
+		case FunctionType::Kind::Error:
+		{
+			solAssert(false, "");
```
Comment (hrkrshnn) [inline libsolidity/codegen/ExpressionCompiler.cpp:917]: Just to confirm, this will be implemented in the follow-up PR, right?

File: libsolidity/ast/ASTUtils.cpp [hunk 9]
```diff
@@ -67,4 +72,5 @@ VariableDeclaration const* rootConstVariableDeclaration(VariableDeclaration cons
 	return rootDecl;
 }
 
+
```
Comment (hrkrshnn) [inline libsolidity/ast/ASTUtils.cpp]:
```suggestion
```

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 8 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:917]: Yes, this is code generation.

File: libsolidity/parsing/Parser.cpp (hunk 7 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp:116]: not sure, it's a rather common term.

File: test/libsolidity/syntaxTests/errors/weird2.sol (hunk 5 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/errors/weird2.sol]: Ah right, I actually wanted to add `selector` to the type.

**Review by hrkrshnn — DISMISSED** [review]
Can be merged after squashing the last commit.

**Comment by chriseth** [comment]
We should discuss on Wednesday if we want to merge this already.

File: docs/grammar/Solidity.g4 [hunk 10]
```diff
@@ -367,7 +381,7 @@ inlineArrayExpression: LBrack (expression ( Comma expression)* ) RBrack;
 /**
  * Besides regular non-keyword Identifiers, the 'from' keyword can also occur as identifier outside of import statements.
```
Comment (cameel) [inline docs/grammar/Solidity.g4]:
```suggestion
 * Besides regular non-keyword Identifiers, some keywords like 'from' or 'error' can also be used as identifiers.
```

File: libsolidity/analysis/DocStringTagParser.cpp [hunk 11]
```diff
@@ -134,11 +141,14 @@ void DocStringTagParser::handleCallable(
 )
 {
 	static set<string> const validEventTags = set<string>{"dev", "notice", "return", "param"};
+	static set<string> const validErrorTags = set<string>{"dev", "notice", "param"};
```
Comment (cameel) [inline libsolidity/analysis/DocStringTagParser.cpp:144]: Just noticed this: why is `@return` valid for events? They don't have return arguments, do they?

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 12]
```diff
@@ -29,6 +29,7 @@
 #include <liblangutil/SemVerHandler.h>
 
 #include <libsolutil/UTF8.h>
+#include <libsolutil/FunctionSelector.h>
```
Comment (cameel) [inline libsolidity/analysis/SyntaxChecker.cpp]: This seems unused. I don't see `selectorFromSignature` anywhere in this file.

File: libsolidity/analysis/ContractLevelChecker.cpp [hunk 13]
```diff
@@ -433,6 +433,24 @@ void ContractLevelChecker::checkHashCollisions(ContractDefinition const& _contra
 			);
 		hashes.insert(hash);
 	}
+
+	map<uint32_t, SourceLocation> errorHashes;
+	for (ErrorDefinition const* error: _contract.interfaceErrors())
+	{
+		if (!error->functionType(true)->interfaceFunctionType())
+			// Will create an error later on.
```
Comment (cameel) [inline libsolidity/analysis/ContractLevelChecker.cpp]:
```suggestion
			// Not having an external type will result in a compilation error later anyway so we can ignore it here.
```

At first I interpreted it as a TODO. I'd reword it to make it clearer:

File: libsolidity/analysis/ContractLevelChecker.cpp [hunk 14]
```diff
@@ -433,6 +433,24 @@ void ContractLevelChecker::checkHashCollisions(ContractDefinition const& _contra
...
+			continue;
+		uint32_t hash = selectorFromSignature32(error->functionType(true)->externalSignature());
+		if (errorHashes.count(hash))
+			m_errorReporter.typeError(
+				4883_error,
+				_contract.location(),
+				SecondarySourceLocation{}.append("This error has the same selector: ", errorHashes[hash]),
+				string("Error signature hash collision for ") + error->functionType(true)->externalSignature()
```
Comment (cameel) [inline libsolidity/analysis/ContractLevelChecker.cpp]:
```suggestion
				"Error signature hash collision for "s + error->functionType(true)->externalSignature()
```

File: libsolidity/analysis/ContractLevelChecker.cpp [hunk 15]
```diff
@@ -433,6 +433,24 @@ void ContractLevelChecker::checkHashCollisions(ContractDefinition const& _contra
...
+	{
+		if (!error->functionType(true)->interfaceFunctionType())
+			// Will create an error later on.
+			continue;
+		uint32_t hash = selectorFromSignature32(error->functionType(true)->externalSignature());
+		if (errorHashes.count(hash))
+			m_errorReporter.typeError(
+				4883_error,
+				_contract.location(),
+				SecondarySourceLocation{}.append("This error has the same selector: ", errorHashes[hash]),
```
Comment (cameel) [inline libsolidity/analysis/ContractLevelChecker.cpp]:
```suggestion
				SecondarySourceLocation{{"This error has the same selector: ", errorHashes[hash]}},
```

File: libsolidity/ast/AST.h [hunk 16]
```diff
@@ -1155,6 +1158,47 @@ class EventDefinition: public CallableDeclaration, public StructurallyDocumented
...
+
+	void accept(ASTVisitor& _visitor) override;
+	void accept(ASTConstVisitor& _visitor) const override;
+
+	TypePointer type() const override;
+
+	FunctionTypePointer functionType(bool _internal) const override;
+
+	bool isVisibleInDerivedContracts() const override { return true; }
+	bool isVisibleViaContractTypeAccess() const override { return true; }
```
Comment (cameel) [inline libsolidity/ast/AST.h:1193]: I see that the same function for events is defined as `{ return false; /* TODO */ }`. What is the `TODO` about? AFAIK `C.Event` works fine so is it about enabling `type(C).Event` or something?

File: libsolidity/ast/AST.h [hunk 17]
```diff
@@ -1155,6 +1158,47 @@ class EventDefinition: public CallableDeclaration, public StructurallyDocumented
...
+
+	ErrorDefinitionAnnotation& annotation() const override;
+
+	CallableDeclaration const& resolveVirtual(
+		ContractDefinition const&,
+		ContractDefinition const*
```
Comment (cameel) [inline libsolidity/ast/AST.h:1199]: I think it's better to repeat the parameter names, even if they are not mandatory in the declaration.

File: libsolidity/ast/AST.cpp [hunk 18]
```diff
@@ -162,6 +162,15 @@ vector<EventDefinition const*> const& ContractDefinition::interfaceEvents() cons
...
 }
 
+vector<ErrorDefinition const*> ContractDefinition::interfaceErrors() const
+{
+	set<ErrorDefinition const*, CompareByID> result;
+	// TODO add all referenced errors
+	for (ContractDefinition const* contract: annotation().linearizedBaseContracts)
+		result += filteredNodes<ErrorDefinition>(contract->m_subNodes);
+	return convertContainer<vector<ErrorDefinition const*>>(move(result));
+}
```
Comment (cameel) [inline libsolidity/ast/AST.cpp:184]: What about the TODO? Is it waiting for error support in the call graph?

File: libsolidity/ast/ASTUtils.cpp [hunk 19]
```diff
@@ -24,24 +24,29 @@
...
+Declaration const* referencedDeclaration(Expression const& _expression)
+{
+	if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_expression))
+		return memberAccess->annotation().referencedDeclaration;
+	else if (auto const* identifier = dynamic_cast<Identifier const*>(&_expression))
+		return identifier->annotation().referencedDeclaration;
+	else
+		return nullptr;
+}
+
```
Comment (cameel) [inline libsolidity/ast/ASTUtils.cpp]: We already have a very [similar helper in `codegen/ir/Common.cpp`](https://github.com/ethereum/solidity/blob/v0.8.2/libsolidity/codegen/ir/Common.cpp#L126-L134)), though that one is specifically for `FunctionDefinition`. Should we maybe unify them or at least put them in the same module? It's pretty trivial though so maybe it's not worth the effort...

File: libsolidity/ast/ASTUtils.cpp [hunk 20]
```diff
@@ -24,24 +24,29 @@
...
-		return nullptr;
-	};
-
 	auto visitor = [&](VariableDeclaration const& _variable, util::CycleDetector<VariableDeclaration>& _cycleDetector, size_t _depth)
 	{
 		solAssert(_depth < 256, "Recursion depth limit reached");
 
-		if (auto referencedVarDecl = referencedDeclaration(_variable.value().get()))
+		if (!_variable.value())
+			return;
```
Comment (cameel) [inline libsolidity/ast/ASTUtils.cpp]: Can this actually happen? Maybe it should be an assert instead?

File: libsolidity/ast/Types.cpp [hunk 21]
```diff
@@ -2736,13 +2736,34 @@ FunctionType::FunctionType(EventDefinition const& _event):
...
+		m_parameterNames.size() == m_parameterTypes.size(),
+		"Parameter names list must match parameter types list!"
+	);
 	solAssert(
-			m_returnParameterNames.size() == m_returnParameterTypes.size(),
-			"Return parameter names list must match return parameter types list!"
-			);
+		m_returnParameterNames.size() == m_returnParameterTypes.size(),
+		"Return parameter names list must match return parameter types list!"
+	);
```
Comment (cameel) [inline libsolidity/ast/Types.cpp]: Shouldn't these lists be empty for events?

File: libsolidity/ast/Types.cpp [hunk 22]
```diff
@@ -3229,6 +3251,8 @@ MemberList::MemberMap FunctionType::nativeMembers(ASTNode const* _scope) const
 		}
 		return {};
 	}
+	case Kind::Error:
+		return {{"selector", TypeProvider::fixedBytes(4)}};
```
Comment (cameel) [inline libsolidity/ast/Types.cpp:3258]: What about events? Don't they have selectors?

File: test/libsolidity/syntaxTests/errors/clash_function_error_inheritance.sol [hunk 23]
```diff
@@ -0,0 +1,4 @@
+contract A { function Err() public pure {} }
+contract B is A { error Err(); }
+// ----
+// DeclarationError 9097: (63-75): Identifier already declared.
```
Comment (cameel) [inline test/libsolidity/syntaxTests/errors/clash_function_error_inheritance.sol:4]: How about inheriting from two contracts where one has an error and the other has a function?

File: test/libsolidity/syntaxTests/errors/weird4.sol [hunk 24]
```diff
@@ -0,0 +1,9 @@
+error E();
```
Comment (cameel) [inline test/libsolidity/syntaxTests/errors/weird4.sol:1]: Some more test ideas:
- Using error outside of a contract where it is defined.
- Using error's selector outside of a contract where it is defined.
- Using `.selector` on an error "instance" (`E().selector`).
- `anonymous` error.
- `using E for L`, `using L for E`.

**Review by cameel — COMMENTED** [review]
I haven't found anything serious. Pretty much only minor/stylistic tweaks and test case suggestions.

The only other thing is that I noticed that we allow `@return` for events and have assertions about their return arguments. I thought events do not have those?

There are also some unresolved TODOs but given that there are more PRs on top of this, they might be fine.

File: libsolidity/ast/AST.cpp (hunk 18 above)
Comment (cameel) [reply libsolidity/ast/AST.cpp:184]: OK, I see that's resolved in a subsequent PR: https://github.com/ethereum/solidity/pull/11037#discussion_r596822074.

File: libsolidity/ast/AST.h (hunk 16 above)
Comment (chriseth) [reply libsolidity/ast/AST.h:1193]: To be honest, I don't know.

File: libsolidity/analysis/ContractLevelChecker.cpp (hunk 15 above)
Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp]: This does not work for me.

File: libsolidity/ast/Types.cpp (hunk 22 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:3258]: Events are a bit different because their selectors are not truncated to 4 bytes.

File: libsolidity/ast/ASTUtils.cpp (hunk 20 above)
Comment (chriseth) [reply libsolidity/ast/ASTUtils.cpp]: For a constant it cannot, true!

File: test/libsolidity/syntaxTests/errors/clash_function_error_inheritance.sol (hunk 23 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/errors/clash_function_error_inheritance.sol:4]: Added another test.

File: libsolidity/ast/Types.cpp (hunk 22 above)
Comment (cameel) [reply libsolidity/ast/Types.cpp:3258]: Oh, that's an interesting bit of info. I was not aware of that.

File: libsolidity/analysis/DocStringTagParser.cpp (hunk 11 above)
Comment (cameel) [reply libsolidity/analysis/DocStringTagParser.cpp:144]: What about this one? Since you changed https://github.com/ethereum/solidity/pull/10987#discussion_r596884364 I think the `return` tag should not be allowed here.

File: libsolidity/analysis/ContractLevelChecker.cpp (hunk 15 above)
Comment (cameel) [reply libsolidity/analysis/ContractLevelChecker.cpp]: Weird. This is the second time I'm suggesting this and this time I remember that I even checked locally to make sure it works.

Looks like I just forgot the extra `{}` (it's a pair in a vector in a struct and I just tested it for a pair in a vector). This one works for me:

```c++
SecondarySourceLocation{{{"This error has the same selector: "s, errorHashes[hash]}}}
```

**Review by cameel — DISMISSED** [review]
I think this is ready now.

There's still one pending question where I'd like to get an answer but it's about events, not errors so there's no reason for it to block this PR: https://github.com/ethereum/solidity/pull/10987#discussion_r596791933

File: libsolidity/analysis/DocStringTagParser.cpp (hunk 11 above)
Comment (chriseth) [reply libsolidity/analysis/DocStringTagParser.cpp:144]: Yes, we should fix it.

Comment (chriseth) [reply libsolidity/analysis/DocStringTagParser.cpp:144]: I added it to https://github.com/ethereum/solidity/issues/11156

**Comment by leonardoalt** [comment]
DO NOT MERGE!
@cameel said that @chriseth wants to merge all custom errors PR together.
